### PR TITLE
fix: enforce latestActivationDate also on Android

### PR DESCRIPTION
**Pull request by toretefre**
Closing https://github.com/AtB-AS/kundevendt/issues/8886

Based on suggestion from https://github.com/AtB-AS/mittatb-app/pull/3886#issuecomment-1731197343

Enforces `latestActivationDate` also on Android where the `maximumDate` prop is not respected in our datepicker.

If a date later than the latest activation date is entered, a yellow warning `MessageBox` will appear and the submit button will be disabled. These measures are reverted when a valid date before the latest activation date is set, or when a different category / product is selected.

https://github.com/AtB-AS/mittatb-app/assets/21310942/ad92f541-c38d-4419-892b-6bc7c3b97ba0

### CODE REVIEW

File: src/stacks-hierarchy/Root_PurchaseOverviewScreen/components/StartTimeSelection.tsx [hunk 1]
```diff
@@ -97,4 +103,7 @@ const useStyles = StyleSheet.createThemeHook((theme) => ({
   radioSegments: {
     marginTop: theme.spacings.medium,
   },
+  warningBox: {
```
Comment (marius-at-atb) [inline src/stacks-hierarchy/Root_PurchaseOverviewScreen/components/StartTimeSelection.tsx:106]: Is this being used?

File: src/stacks-hierarchy/Root_PurchaseOverviewScreen/components/TravelDate/TravelDateSheet.tsx [hunk 2]
```diff
@@ -29,15 +30,53 @@ type Props = {
...
+  ) => {
     const {t, language} = useTranslation();
     const styles = useStyles();
 
     const defaultDate = travelDate ?? new Date().toISOString();
     const [dateString, setDate] = useState(defaultDate);
+    const [
+      replicatedShowActivationDateWarning,
+      setReplicatedShowActivationDateWarning,
+    ] = useState<boolean>(false);
```
Comment (marius-at-atb) [inline src/stacks-hierarchy/Root_PurchaseOverviewScreen/components/TravelDate/TravelDateSheet.tsx:57]: Should init to showActivationDateWarning I think

File: src/stacks-hierarchy/Root_PurchaseOverviewScreen/components/TravelDate/TravelDateSheet.tsx [hunk 3]
```diff
@@ -29,15 +30,53 @@ type Props = {
...
+    };
+
+    const onSetDate = (date: string) => {
+      if (!maximumDate) setDate(date);
+      else {
+        if (isAfter(date, maximumDate)) {
+          setInternalAndExternalWarningState(true);
+        } else if (
+          replicatedShowActivationDateWarning ||
+          showActivationDateWarning
```
Comment (marius-at-atb) [inline src/stacks-hierarchy/Root_PurchaseOverviewScreen/components/TravelDate/TravelDateSheet.tsx:71]: When replicatedShowActivationDateWarning is initted to showActivationDateWarning, `|| showActivationDateWarning` should not be needed here

**Review by marius-at-atb — APPROVED** [review]
Very nice! The comments are just some nitpicks.
I agree with you [here](https://github.com/AtB-AS/mittatb-app/pull/3886#issuecomment-1731341612) that the state replication is probably not best practice. At the very least maybe make it in a generalised reusable way for all components opened with openBottomSheet. However, for now it is an easy enough solution that works I guess..

File: src/stacks-hierarchy/Root_PurchaseOverviewScreen/components/StartTimeSelection.tsx (hunk 1 above)
Comment (toretefre) [reply src/stacks-hierarchy/Root_PurchaseOverviewScreen/components/StartTimeSelection.tsx:106]: Nope, there was a message for testing, but removed it. Good catch 😄

File: src/stacks-hierarchy/Root_PurchaseOverviewScreen/components/TravelDate/TravelDateSheet.tsx (hunk 2 above)
Comment (toretefre) [reply src/stacks-hierarchy/Root_PurchaseOverviewScreen/components/TravelDate/TravelDateSheet.tsx:57]: Smart!

**Comment by toretefre** [comment]
> Very nice! The comments are just some nitpicks. I agree with you [here](https://github.com/AtB-AS/mittatb-app/pull/3886#issuecomment-1731341612) that the state replication is probably not best practice. At the very least maybe make it in a generalised reusable way for all components opened with openBottomSheet. However, for now it is an easy enough solution that works I guess..

Yep, but I think it makes more sense to make it general when we have another use case. 

For now, merging as-is. Thanks for the help! 🙌

**Review by gorandalum — COMMENTED** [review]
Just a general comment from me, and not something that needs to be changed in a PR :)

My intuitive thought is that the `showActivationDateWarning` is really derived state, based on the selected date and the latest activation date, and as such doesn't need to be its own state, neither in the sheet nor in the purchase overview.

Quote from [the doc](https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state):
> When something can be calculated from the existing props or state, [don’t put it in state.](https://react.dev/learn/choosing-the-state-structure#avoid-redundant-state) Instead, calculate it during rendering.

And I also know that there is a difference between theory and practice, so it might still be needed here, even though I don't see it on first glance 😊

**Pull request merged**